### ISSUE TITLE
Object detection server?

### ISSUE THREAD

**Issue by corot** [GT-RAIL/rail_mesh_icp#4]
Question: did you develop, or do yo plan to do an object detection server?
With current services you can check the fitting of a segmentated pointcloud (extracted with rail_segmentation) to a know mesh. But would make sense to provide an object detection action server that fits each segmentated pointcloud to all the known objects and assigns the best fit (if above a threshold)? That's what the old [ORK tabletop](https://wg-perception.github.io/ork_tutorials/tutorial02/tutorial.html) did, and was very convenient, but current code doesn't work on melodic. I'm surprise that I cannot find something similar.

**Comment by adaruna3** [comment]
Hi Jorge,

The pipeline does not have an object detection server built in. However, if you look at 6.2 in the Wiki, you can see that you can integrate the pipeline with an object detector. Specifically, the `initial_estimate` argument for the `match_template` service calls can be provided and updated. The argument could be provided by an object detector, as you had in mind, so that the initial estimate can change with each call.

Note that although such as object detector might detect multiple objects, each instance of `template_matcher_node` is only matching to a single model (source) pointcloud. Therefore, to get the full functionality you mentioned, you would need to have a `template_matcher_node` instance for each object type. 

I will improve the code so that you can match a different model (source) pointcloud to the target pointcloud with each service call, removing the need for multiple instances of `template_matcher_node`. However, I will need to consider whether to include a full object detection server, as such a server remains out-of-scope of a mesh matching package.

I will close once I incorporate the functionality I mentioned.

Thanks!
Angel

**Comment by corot** [comment]
> The pipeline does not have an object detection server built in. However, if you look at 6.2 in the Wiki, you can see that you can integrate the pipeline with an object detector. Specifically, the `initial_estimate` argument for the `match_template` service calls can be provided and updated. The argument could be provided by an object detector, as you had in mind, so that the initial estimate can change with each call.

Yep, that's what I had in mind;   but as u said, very awkward to have so many template matchers running!
 
> Note that although such as object detector might detect multiple objects, each instance of `template_matcher_node` is only matching to a single model (source) pointcloud. Therefore, to get the full functionality you mentioned, you would need to have a `template_matcher_node` instance for each object type.
> 
> I will improve the code so that you can match a different model (source) pointcloud to the target pointcloud with each service call, removing the need for multiple instances of `template_matcher_node`. However, I will need to consider whether to include a full object detection server, as such a server remains out-of-scope of a mesh matching package.

Yes, something like loading all the templates on in input directory and label the objects after the file name w/o extension would be great. I volunteer as beta-tester! :raised_hand_with_fingers_splayed: :smile: 

Thanks!!!

**Comment by corot** [comment]
Merry Christmas!  Is there any work in progress on the multiple templates loading?

**Comment by corot** [comment]
Btw, I modified ICP to only allow planar transformation (for tabletop objects) by using a TransformationEstimation2D. 
Would it make sense to add an option to the ICPMatch service to indicate the possible transformations? (2D, 3D, 6D...)
I can open a PR if so
